x/ibc/03-connection/types/version.go

```diff
@@ -1,96 +1,171 @@
 package types
 
-// GetCompatibleVersions returns an ordered set of compatible IBC versions for the
-// caller chain's connection end.
+import (
+	"fmt"
+	"strings"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+)
+
+var (
+	// DefaultIBCVersion represents the latest supported version of IBC.
+	// The current version supports only ORDERED and UNORDERED channels.
+	DefaultIBCVersion = CreateVersionString("1", []string{"ORDERED channel", "UNORDERED channel"})
+)
+
+// GetCompatibleVersions returns a descending ordered set of compatible IBC
+// versions for the caller chain's connection end. The latest supported
+// version should be first element and the set should descend to the oldest
+// supported version.
 func GetCompatibleVersions() []string {
-	return []string{"1.0.0"}
+	return []string{DefaultIBCVersion}
 }
 
-// LatestVersion gets the latest version of a connection protocol
+// CreateVersionString constructs a valid connection version given a
+// version identifier and feature set. The format is written as:
 //
-// CONTRACT: version array MUST be already sorted.
-func LatestVersion(versions []string) string {
-	if len(versions) == 0 {
-		return ""
-	}
-	return versions[len(versions)-1]
+// ([version_identifier],[feature_0,feature_1,feature_2...])
+//
+// A connection version is considered invalid if it is not in this format
+// or violates one of these rules:
+// - the version identifier is empty or contains commas
+// - a specified feature contains commas
+func CreateVersionString(identifier string, featureSet []string) string {
+	return fmt.Sprintf("(%s,%s)", identifier, featureSet)
 }
 
-// PickVersion picks the counterparty latest version that is matches the list
-// of compatible versions for the connection.
-func PickVersion(counterpartyVersions, compatibleVersions []string) string {
+// UnpackVersion parses a version string and returns the identifier and the
+// feature set of a version. An error is returned if the version is not valid.
+func UnpackVersion(version string) (string, []string, error) {
+	// validate version so valid splitting assumptions can be made
+	if err := host.ConnectionVersionValidator(version); err != nil {
+		return "", nil, err
+	}
 
-	n := len(counterpartyVersions)
-	m := len(compatibleVersions)
+	// peel off prefix and suffix of the tuple
+	version = strings.TrimPrefix(version, "(")
```

Comment:
... ah I see. This is alright, we just need to ban the character `(` in the first part of the string.

Maybe we should use `proto` here instead of this encoding?

Reply:
this should be handled by the regex validation. I left out these sorts of checks in the `CreateVersionString` func because I made the assumption that versions would always be validated afterwards anyways and not returning an error on create would be cleaner, but I can update.

With regards to `proto` encoding, would this mean passing in proto encoded version as a string and then decoding it to a version struct? Or is there a way to parse with proto? I definitely think using some sort of encoding/decoding would be cleaner with the code. My only concern here is imposing the version string to be encoded with proto would require counterparties to abide by this requirement. But that is probably insignificant. Also I guess the version set in state would be proto encoded?

Reply:
this might be something better done by someone more familiar with `proto` than me in a followup pr. I'm going to finish up this pr under this assumption. I don't feel comfortable in designing a solution that doesn't involve entangling dependencies on some centralized proto struct

Reply:
I suppose we could change it in the actual handshake datagrams, having a layer of abstraction may not buy us much flexibility here since the two sides have to understand each other's versions anyways. This is alright for now I think.